File: new/python/150_evaluate_reverse_polish_notation_doubly_linked_list.py
class DoublyLinkedNode:
    def __init__(self, val: int, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        # idea: doubly linked node
        # create a doubly linked list out of the given tokens by iterating through each token in the tokens and making each token
        # a doubly linked node. connecting all the tokens to form a doubly linked list.
        # once we have doubly linked list, while loop the linked list and find the current node's value is a symbol "+-*/",
        # if it is a symbol, get a and b, left->left->val and left->val, and get the result for the symbol.
        # store the result of the value in the current node, and make current node's left point to left->left->left and then
        # now current node's left point to left->left->left, and make the current node's left point to the current node
        # (original left->left->left's right point to current node, just simply current->left->right = current node).
        # move the current to the next doubly linked node.
        head = DoublyLinkedNode(tokens[0], left=None, right=None)
        curr = head

        # make a doubly linked list
        for i in range(1, len(tokens)):
            curr.right = DoublyLinkedNode(tokens[i], left=curr, right=None)
            curr = curr.right

        curr = head
        while curr is not None:
            if curr.val in "+-*/":
                a = int(curr.left.left.val)
                b = int(curr.left.val)
                if curr.val == "+":
                    result = a + b
                elif curr.val == "-":
                    result = a - b
                elif curr.val == "*":
                    result = a * b
                else:
                    result = int(a / b)
                curr.val = str(result)

                curr.left = curr.left.left.left
                if curr.left is not None:
                    curr.left.right = curr  # successfully cut off the nodes that have values of a and b.

            value = curr.val
            curr = curr.right

        return int(value)
        # time: O(n)
        # space: O(n)

File: new/cpp/150_evaluate_reverse_polish_notation_doubly_linked_list.cpp
class DoublyLinkedNode {
    public:
        string val;
        DoublyLinkedNode* prev;
        DoublyLinkedNode* next;

        DoublyLinkedNode(string val, DoublyLinkedNode* prev = nullptr, DoublyLinkedNode* next = nullptr) {
            this->val = val;
            this->prev = prev;
            this->next = next;
        }
};


class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        // idea: doubly linked node
        // create a doubly linked list out of the given tokens
        // in the doubly linked list, loop until we find these symbols: "+-*/"
        // if we find these symbols, then two positions previous to the symbol is "a"
        // and 1 position prev to the symbol is "b". We do the given symbol's operation with a and b in that order, and find the result
        // and store the result in the current node's value. 
        // now that we used prev->prev and prev, we don't need it anymore, so we update our curren't previous to prev->prev->prev, and 
        // prev->prev->prev 's next to our current node, essentially eliminating the middle numbers (a and b). 
        DoublyLinkedNode* head = new DoublyLinkedNode(tokens[0], nullptr, nullptr);
        DoublyLinkedNode* curr = head;
        
        // make a linked list out of the tokens. 
        for (int i = 1; i < tokens.size(); i++) {
            curr->next = new DoublyLinkedNode(tokens[i], curr, nullptr);
            curr = curr->next;
        }

        int output = 0;
        curr = head; // !!!!!!! otherwise use head, or different pointer. Forgot this !!!!
        while (curr != nullptr) {

            if (curr->val == "+" || curr->val == "-" || curr->val == "*" || curr->val == "/") {
                int a = stoi(curr->prev->prev->val);
                int b = stoi(curr->prev->val);
                int result = 0;
                if (curr->val == "+") {
                    result = a + b;
                } else if (curr->val == "-") {
                    result = a - b;
                } else if (curr->val == "*") {
                    result = a * b;
                } else {
                    result = a / b;
                }

                curr->val = to_string(result);
                curr->prev = curr->prev->prev->prev;
                if (curr->prev != nullptr) {
                    curr->prev->next = curr;
                }
            }

            output = stoi(curr->val);
            curr = curr->next;
        }

        return output;
    }
    // time: O(n)
    // space: O(n)
};
